PopPUNK/lineage_clustering.py

```diff
@@ -0,0 +1,680 @@
+                                existing_scheme = existing_scheme,
+                                distances_length = distances.shape,
+                                distances_type = distances.dtype),
+                                rank_list)
+        
+        for n,result in enumerate(results):
+            R = rank_list[n]
+            lineage_clustering[R], lineage_seed[R], neighbours[R], previous_lineage_clustering[R] = result
+        
+    # manage memory
+    shm_distances.close()
+    shm_distances.unlink()
+    del shm_distances
+    shm_distance_ranks.close()
+    shm_distance_ranks.unlink()
+    del shm_distance_ranks
+    isolate_list_shared.shm.close()
+    isolate_list_shared.shm.unlink()
+    del isolate_list_shared
+
+    # store output
+    with open(output + "/" + output + '_lineageClusters.pkl', 'wb') as pickle_file:
+        pickle.dump([lineage_clustering, lineage_seed, neighbours, rank_list], pickle_file)
+    
+    # print output
+    combined = {}
+    titles_list = ['Lineage_R' + str(R)  for R in rank_list]
+    lineage_output_name = output + "/" + output + "_lineage_clusters.csv"
+    with open(lineage_output_name, 'w') as lFile:
+        # print header
+        lFile.write('Id')
+        for t in titles_list:
+            lFile.write(',' + t + '__autocolor')
+            combined[t] = {}
+        lFile.write(',Overall_lineage')
+        combined['Overall_lineage'] = {}
+        if qlist is not None:
+            lFile.write(',Status')
+        lFile.write('\n')
+
+        # print lines for each isolate
+        for isolate in lineage_clustering[R].keys():
+            lFile.write(isolate)
+            for n,R in enumerate(rank_list):
+                lFile.write(',' + str(lineage_clustering[R][isolate]))
+                lineage_string = str(lineage_clustering[R][isolate])
+                # include information on lineage clustering
+                combined[titles_list[n]][isolate] = lineage_string
+                if lineage_clustering[R][isolate] != previous_lineage_clustering[R][isolate] and previous_lineage_clustering[R][isolate] != null_cluster_value:
+                    lineage_string = str(previous_lineage_clustering[R][isolate]) + ':' + lineage_string
+                if isolate in combined['Overall_lineage'].keys():
+                    combined['Overall_lineage'][isolate] = combined['Overall_lineage'][isolate] + '-' + lineage_string
+                else:
+                    combined['Overall_lineage'][isolate] = lineage_string
+            lFile.write(',' + combined['Overall_lineage'][isolate])
+            if qlist is not None:
+                if isolate in qlist:
+                    lFile.write(',Added')
+                else:
+                    lFile.write(',Existing')
+            lFile.write('\n')
+
+    return combined
+
+def run_clustering_for_R(R, null_cluster_value = None, qlist = None, existing_scheme = False, distances_length = None, distances_type = None):
```

Comment:
R is a bad variable name

Reply:
rank is probably better

Reply:
Aware we also use X throughout, so probably being inconsistent with criticism

Reply:
Changed in 01cd435a731ae8cbc268fe7eacc4ea0dc7f42eb9.